File: ui/src/components/agents/validateAgent.ts
import type { AgentDefinitionEntry } from '../../ipc/protocol.js'
import type { AgentSource, AgentValidationResult, DraftAgent } from './types.js'
import { getAgentSourceDisplayName } from './utils.js'

/**
 * Lite-native port of upstream's
 * `ui/examples/upstream-patterns/src/components/agents/validateAgent.ts`.
 *
 * Upstream validated against an in-process `Tools` registry; the Lite
 * frontend can't walk the backend's tool registry synchronously, so
 * the tool validity check is simplified — callers can pass an
 * `availableTools` list (the IPC payload provides one) and we flag any
 * missing names as invalid.
 */

export function validateAgentType(agentType: string): string | null {
  if (!agentType) {
    return 'Agent type is required'
  }
  if (!/^[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]$/.test(agentType)) {
    return 'Agent type must start and end with alphanumeric characters and contain only letters, numbers, and hyphens'
  }
  if (agentType.length < 3) {
    return 'Agent type must be at least 3 characters long'
  }
  if (agentType.length > 50) {
    return 'Agent type must be less than 50 characters'
  }
  return null
}

export function validateAgent(
  agent: DraftAgent,
  availableTools: string[],
  existingAgents: AgentDefinitionEntry[],
): AgentValidationResult {
  const errors: string[] = []
  const warnings: string[] = []

  if (!agent.agentType) {
    errors.push('Agent type is required')
  } else {
    const typeError = validateAgentType(agent.agentType)
    if (typeError) errors.push(typeError)

    const duplicate = existingAgents.find(
      a => a.name === agent.agentType && readSource(a) !== agent.source,
    )
    if (duplicate) {
      errors.push(
        `Agent type "${agent.agentType}" already exists in ${getAgentSourceDisplayName(
          readSource(duplicate) as AgentSource,
        )}`,
      )
    }
  }

  if (!agent.description) {
    errors.push('Description is required')
  } else if (agent.description.length < 10) {
    warnings.push('Description should be more descriptive (at least 10 characters)')
  } else if (agent.description.length > 5000) {
    warnings.push('Description is very long (over 5000 characters)')
  }

  if (agent.tools !== undefined && !Array.isArray(agent.tools)) {
    errors.push('Tools must be an array')
  } else if (agent.tools === undefined) {
    warnings.push('Agent has access to all tools')
  } else if (agent.tools.length === 0) {
    warnings.push('No tools selected — agent will have very limited capabilities')
  } else {
    const availableSet = new Set(availableTools)
    const invalid = agent.tools.filter(t => !availableSet.has(t))
    if (invalid.length > 0) {
      errors.push(`Invalid tools: ${invalid.join(', ')}`)
    }
  }

  if (!agent.systemPrompt) {
    errors.push('System prompt is required')
  } else if (agent.systemPrompt.length < 20) {
    errors.push('System prompt is too short (minimum 20 characters)')
  } else if (agent.systemPrompt.length > 10000) {
    warnings.push('System prompt is very long (over 10,000 characters)')
  }

  return {
    isValid: errors.length === 0,
    errors,
    warnings,
  }
}

function readSource(entry: AgentDefinitionEntry): string {
  return entry.source?.kind ?? 'unknown'
}

File: crates/claude-code-rs/src/ui/agents/validate_agent.rs
//! Validation helpers for editable agent definitions.

use super::types::AgentValidationResult;

pub fn validate_agent_definition(
    agent_type: &str,
    when_to_use: &str,
    system_prompt: &str,
    existing_agent_types: &[String],
) -> AgentValidationResult {
    let mut result = AgentValidationResult::ok();

    let name = agent_type.trim();
    if name.is_empty() {
        result = result.with_error("Agent name is required");
    } else {
        if name.len() > 64 {
            result = result.with_error("Agent name must be 64 characters or fewer");
        }
        if !name
            .chars()
            .all(|ch| ch.is_ascii_alphanumeric() || matches!(ch, '-' | '_'))
        {
            result = result.with_error("Agent name may contain only letters, numbers, '-' and '_'");
        }
        if existing_agent_types
            .iter()
            .any(|existing| existing.eq_ignore_ascii_case(name))
        {
            result = result.with_error("Agent name already exists");
        }
    }

    if when_to_use.trim().is_empty() {
        result = result.with_error("Description is required");
    } else if when_to_use.trim().len() < 12 {
        result = result.with_warning("Description is very short");
    }

    if system_prompt.trim().is_empty() {
        result = result.with_error("System prompt is required");
    } else if system_prompt.trim().len() < 40 {
        result = result.with_warning("System prompt may be too short for reliable delegation");
    }

    result
}

pub fn render_validation_result(result: &AgentValidationResult) -> String {
    let mut lines = vec![format!(
        "status: {}",
        if result.is_valid { "valid" } else { "invalid" }
    )];
    for error in &result.errors {
        lines.push(format!("error: {error}"));
    }
    for warning in &result.warnings {
        lines.push(format!("warning: {warning}"));
    }
    lines.join("\n")
}
